### PR TITLE
add test and predict

### PR DESCRIPTION
**Pull request by shippingwang**
This PR
- add predict function, which directly calls paddle built-in predictor to predict.
- add test function

for now,  the panorama of PaddleVideo, function level.

| func| build model |backward | loss | metric(top1/5)|predict ret|
|- |:-: |:-: |:-: |:-: |:-:|
|Train|Y|Y|Y|Y||
|Val   |Y|   |Y|Y||
|Test |Y|   |  |Y||
 |Infer(not impl) |Y|   |  |  |Y|
| Predict(py)|||  |  |Y|

- Others:
  - add save_interval and output path to refine save function.
  - fix bugs when saving the best model.
  - refactoring forward to implement inferencing

image

Because of predictor will sign the forward method as the total graph, and in the old design of PaddleVideo we integrate build_model and build_loss together, so the incorrect results occur, we create new interfaces (train_step, val_step, test_step) to directly call forward_train/val and indirectly call forward to decouple build_loss and build_model now

Todo:
- check shared memory problem
- benchmark

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -18,35 +18,32 @@
...
     parser.add_argument(
         '--validate',
         action='store_true',
         help='whether to evaluate the checkpoint during training')
-    parser.add_argument(
-        '--seed',
-        type=int,
-        default=None,
-        help='random seed')
+    parser.add_argument('--seed', type=int, default=None, help='random seed')
```
Comment (huangjun12) [inline main.py:39]: seed not used?

File: paddlevideo/modeling/framework/recognizers/base.py [hunk 2]
```diff
@@ -50,25 +47,31 @@ def extract_feature(self, imgs):
...
+        batches = imgs.shape[0]
+        num_segs = imgs.shape[1]
+        imgs = paddle.reshape(imgs, [-1] + list(imgs.shape[2:]))
+
+        feature = self.extract_feature(imgs)
+        cls_score = self.head(feature, num_segs)
+
+        return cls_score
 
     def train_step(self, data_batch, **kwargs):
```
Comment (huangjun12) [inline paddlevideo/modeling/framework/recognizers/base.py:76]: Is train_step and val_step necessary when we use forward_train and forward_test?

File: paddlevideo/tasks/train.py [hunk 3]
```diff
@@ -120,29 +126,40 @@ def evaluate():
...
+                best = evaluate(best)
+
+            # save best
+            save(opt_state_dict, osp.join(output_dir, f"{opt_name}.pdopt"))
+            save(model.state_dict(),
+                 osp.join(output_dir, model_name + "_best.pdparams"))
+            best = int(best * 10000) / 10000
+            logger.info(f"Already save the best model (top1 acc){best}")
+
+        if not validate and epoch % cfg.get("save_interval", 10) == 0:
```
Comment (huangjun12) [inline paddlevideo/tasks/train.py:160]: save的时间:
(1) epoch % save_interval == 0；
(2) 最后一个epoch
(3) 如果做validate，并且validate的精度与评估精度等价时，保存best。
做了validate就不保存这个条件要改一下？

File: paddlevideo/tasks/train.py [hunk 4]
```diff
@@ -120,29 +126,40 @@ def evaluate():
...
+        opt_state_dict = optimizer.state_dict()
+        opt_name = cfg['OPTIMIZER']['name']
+
         if cfg.get("PRECISEBN") and (epoch -
                                      1) % cfg.PRECISEBN.preciseBN_interval == 0:
             do_preciseBN(
                 model, train_loader, parallel,
                 min(cfg.PRECISEBN.num_iters_preciseBN, len(train_loader)))
-
         if validate:
```
Comment (huangjun12) [inline paddlevideo/tasks/train.py:149]: 设置一下validate间隔？

File: paddlevideo/tasks/test.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# copyright (c) 2020 PaddlePaddle Authors. All Rights Reserve.
```
Comment (huangjun12) [inline paddlevideo/tasks/test.py:1]: 这个文件名也叫test? 跟tool/test下的文件名重了

File: paddlevideo/tasks/train.py (hunk 3 above)
Comment (shippingwang) [reply paddlevideo/tasks/train.py:160]: save_interval 必保存， best是 val可选的，最后一个epoch必须保存

File: paddlevideo/modeling/framework/recognizers/base.py (hunk 2 above)
Comment (shippingwang) [reply paddlevideo/modeling/framework/recognizers/base.py:76]: would refactor in future
